File: live/sortviewer/code/Sort.js
function swapValue(arr, i, j) {
    var t = arr[i];
    arr[i] = arr[j];
    arr[j] = t;
}

function bubbleSort(arr) {
    var indexOfLastUnsortedElement = arr.length;
    var swapped = true;
    while (swapped) {
        swapped = false;
        for (var i = 0; i < indexOfLastUnsortedElement - 1; i++)
            if (arr[i] > arr[i + 1]) {
                swapValue(arr, i, i + 1);
                swapped = true;
            }
        indexOfLastUnsortedElement--;
    }
}

function selectionSort(arr) {
    for (var i = 0; i < arr.length - 1; i++) {
        var minIdx = i;
        for (var j = i + 1; j < arr.length; j++)
            if (arr[j] < arr[minIdx]) minIdx = j;
        if (i < minIdx) swapValue(arr, i, minIdx)
    }
}

function insertionSort(arr) {
    for (var i = 1; i < arr.length; ++i) {
        var tempValue = arr[i];
        for (var j = i - 1; j >= 0 && arr[j] > tempValue; --j)
            swapValue(arr, j, j + 1);
        arr[j + 1] = tempValue;
    }
}

function quickSort(arr, left, right) {
    var pivotidx = ((left + right) / 2).toFixed();
    var i = left;
    var j = right;
    var pivot = arr[pivotidx];
    while (i <= j) {
        while (arr[i] < pivot) i++;
        while (arr[j] > pivot) j--;
        if (i <= j) {
            if (arr[i] > arr[j] && i < j)
                swapValue(arr, i, j);
            i++;
            j--;
        }
    }
    if (left < j) quickSort(arr, left, j);
    if (i < right) quickSort(arr, i, right);
    return arr;
}

function mergeSort(arr, left, right) {
    if (right > left) {
        var mid = left + Math.floor((right - left) / 2);
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        var subLeft = [];
        var subRight = [];
        for (var i = left; i <= mid; i++) subLeft.push(arr[i]);
        for (var i = mid + 1; i <= right; i++) subRight.push(arr[i]);
        var leftIndex = 0;
        var rightIndex = 0;
        for (var i = left; i <= right; i++)
            if (left + leftIndex <= mid && (mid + 1 + rightIndex > right || subLeft[leftIndex] < subRight[rightIndex])) {
                arr[i] = subLeft[leftIndex];
                leftIndex++;
            } else {
                arr[i] = subRight[rightIndex];
                rightIndex++;
            }
    }
}

var arr = [6, 5, 3, 1, 5, 7, 2];
bubbleSort(arr);
console.log(arr);
arr = [6, 5, 3, 1, 5, 7, 2];
selectionSort(arr);
console.log(arr);
arr = [6, 5, 3, 1, 5, 7, 2];
insertionSort(arr);
console.log(arr);
arr = [6, 5, 3, 1, 5, 7, 2];
quickSort(arr, 0, arr.length - 1);
console.log(arr);
arr = [6, 5, 3, 1, 5, 7, 2];
mergeSort(arr, 0, arr.length - 1);
console.log(arr);

File: live/sortviewer/code/Sort.pas
type arrayType = array[1..100] of integer;
var myArr: arrayType;
    len, i: byte;

Procedure swapValue(Var arr: arrayType; i, j: byte);
var t: integer;
Begin
    t:= arr[i];
    arr[i]:= arr[j];
    arr[j]:= t;
End;

Procedure bubbleSort(Var arr: arrayType; arrLength: byte);
var lastUnsort, i: byte;
    swapped: boolean;
Begin
    lastUnsort:= arrLength;
    swapped:= true;
    while swapped do
    Begin
        swapped:= false;
        for i:= 1 to lastUnsort - 1 do
            if arr[i] > arr[i + 1] then
            Begin
                swapValue(arr, i, i + 1);
                swapped:= true;
            End;
        lastUnsort:= lastUnsort - 1;
    End;
End;

Procedure selectionSort(Var arr: arrayType; arrLength: byte);
var i, j, minIdx: byte;
Begin
    for i:= 1 to arrLength - 1 do
    Begin
        minIdx:= i;
        for j:= i + 1 to arrLength do
            if arr[j] < arr[minIdx] then minIdx:= j;
        if i < minIdx then swapValue(arr, i, minIdx);
    End;
End;

Procedure insertionSort(Var arr: arrayType; arrLength: byte);
var i, j: byte;
    tempValue: integer;
Begin
    j:= 0;
    for i:= 2 to arrLength do
    Begin
        tempValue:= arr[i];
        for j:= i - 1 downto 1 do
            if arr[j] > tempValue then
                swapValue(arr, j, j + 1)
            else break;
        arr[j + 1]:= tempValue;
    End;
End;

Procedure quickSort (Var arr: arrayType; left, right: byte);
var pivotidx, i, j: byte;
    pivot: integer;
Begin
    pivotidx:= (left + right) div 2;
    i:= left;
    j:= right;
    pivot:= arr[pivotidx];
    //partition
    while i <= j do
    Begin
        while arr[i] < pivot do i:= i + 1;
        while arr[j] > pivot do j:= j - 1;
        if i <= j then
        Begin
            if (arr[i] > arr[j]) and (i < j) then swapValue(arr, i, j);
            i:= i + 1;
            j:= j - 1;
        End;
    End;
    // recursion
    if left < j then quickSort(arr, left, j);
    if i < right then quickSort(arr, i, right);
End;

Procedure mergeSort (Var arr: arrayType; left, right: byte);
Var mid, i, leftIndex, rightIndex : byte;
    subLeft, subRight : arrayType;
Begin
    If right > left Then
    Begin         
        mid:= (left + right) Div 2;
        mergeSort (arr, left, mid);
        mergeSort (arr, mid + 1, right);
        For i:= left to mid Do subLeft[i - left + 1]:= arr[i];
        For i:= mid + 1 To right Do subRight[i - mid]:= arr[i];
        leftIndex := 1;
        rightIndex := 1;
        for i:= left to right do
            if (left + leftIndex - 1 <= mid) and ((mid + rightIndex > right) 
                or (subLeft[leftIndex] < subRight[rightIndex])) then
            Begin
                arr[i]:= subLeft[leftIndex];
                leftIndex:= leftIndex + 1;
            End else Begin
                arr[i]:= subRight[rightIndex];
                rightIndex:= rightIndex + 1;
            End;
    End;
End;

Begin
    Write('Nhap so luong phan tu: '); readln(len);
    for i:= 1 to len do
    Begin
        Write('[', i, '] = '); readln(myArr[i]);
    End;
    Writeln('Mang da xep: ');
    bubbleSort(myArr, len);
    for i:=1 to len do Write(myArr[i], ' '); Writeln;
    selectionSort(myArr, len);
    for i:=1 to len do Write(myArr[i], ' '); Writeln;
    insertionSort(myArr, len);
    for i:=1 to len do Write(myArr[i], ' '); Writeln;
    quickSort(myArr,1, len);
    for i:=1 to len do Write(myArr[i], ' '); Writeln;
    mergeSort(myArr, 1, len);
    for i:=1 to len do Write(myArr[i], ' '); Writeln;
End.